velox/dwio/parquet/reader/IAAPageReader.cpp

```diff
@@ -0,0 +1,508 @@
+    updateRowInfoAfterPageSkipped();
+  }
+  return false;
+}
+
+void IAAPageReader::prefetchDataPageV1(const thrift::PageHeader& pageHeader) {
+  dataPageHeader_ = pageHeader;
+  VELOX_CHECK(
+      pageHeader.type == thrift::PageType::DATA_PAGE &&
+      pageHeader.__isset.data_page_header);
+
+  dataPageData_ = readBytes(pageHeader.compressed_page_size, pageBuffer_);
+  pre_decompress_data = iaaDecompress(
+      dataPageData_,
+      pageHeader.compressed_page_size,
+      pageHeader.uncompressed_page_size,
+      uncompressedDataV1Data_,
+      dataDecompFuture);
+  return;
+}
+
+void IAAPageReader::prefetchDataPageV2(const thrift::PageHeader& pageHeader) {
+  return;
+}
+
+void IAAPageReader::prefetchDictionary(const thrift::PageHeader& pageHeader) {
+  dictPageHeader_ = pageHeader;
+  dictionaryEncoding_ = pageHeader.dictionary_page_header.encoding;
+  VELOX_CHECK(
+      dictionaryEncoding_ == Encoding::PLAIN_DICTIONARY ||
+      dictionaryEncoding_ == Encoding::PLAIN);
+  dictPageData_ = readBytes(pageHeader.compressed_page_size, pageBuffer_);
+
+  pre_decompress_dict = iaaDecompress(
+      dictPageData_,
+      pageHeader.compressed_page_size,
+      pageHeader.uncompressed_page_size,
+      uncompressedDictData_,
+      dictDecompFuture);
+
+  return;
+}
+
+const bool IAAPageReader::iaaDecompress(
+    const char* pageData,
+    uint32_t compressedSize,
+    uint32_t uncompressedSize,
+    BufferPtr& uncompressedData,
+    folly::SemiFuture<uint64_t>& future) {
+  dwio::common::ensureCapacity<char>(
+      uncompressedData, uncompressedSize, &pool_);
+  static constexpr int PARQUET_ZLIB_WINDOW_BITS_4KB = 12;
+  future = folly::makeSemiFuture((uint64_t)0);
+  if (!isWinSizeFit) {
+    // window size should be 4KB for IAA
+    if (PARQUET_ZLIB_WINDOW_BITS_4KB ==
+        dwio::common::compression::getZlibWindowBits(
+            (const uint8_t*)pageData, uncompressedSize)) {
+      isWinSizeFit = true;
+    } else {
+      future = folly::makeSemiFuture((uint64_t)0);
+      return true;
+    }
+  }
+  std::unique_ptr<dwio::common::compression::AsyncDecompressor> decompressor =
+      dwio::common::compression::createAsyncDecompressor(
+          thriftCodecToCompressionKind());
+  if (decompressor == nullptr) {
+    return true;
+  }
+  auto decompFuture = decompressor->decompressAsync(
+      (const char*)pageData,
+      compressedSize,
+      (char*)uncompressedData->asMutable<char>(),
+      uncompressedSize);
+  if (decompFuture.isReady()) {
+    auto result = std::move(decompFuture).getTry();
+    if (result.hasException()) {
+      future = folly::makeSemiFuture((uint64_t)0);
+      return true;
+    }
+  }
+  future = std::move(decompFuture);
+  return true;
+}
+
+void IAAPageReader::seekToPage(int64_t row) {
+  this->defineDecoder_.reset();
+  this->repeatDecoder_.reset();
+  // 'rowOfPage_' is the row number of the first row of the next page.
+  this->rowOfPage_ += this->numRowsInPage_;
+
+  if (seekToPreDecompPage(row)) {
```

Comment:
Maybe create some virtual hooks in the base class for these calls, then you don't need to duplicate the other part in subclasses.